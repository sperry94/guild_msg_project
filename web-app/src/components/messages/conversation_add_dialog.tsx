import React from "react"
import { Formik, Form } from "formik"
import * as yup from "yup"
import { makeStyles } from "@material-ui/core/styles"
import {
  Button,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Typography,
  TextField,
  CircularProgress,
} from "@material-ui/core"
import { addConversation } from "../../services/conversations"

const useStyles = makeStyles({
  dialog: {
    minWidth: "600px",
  },
  input: {
    marginTop: "24px",
    width: "100%",
  },
  inputActions: {
    paddingBottom: 0,
    paddingRight: 0,
    marginBottom: "16px",
    marginRight: "24px",
  },
  spinner: {
    color: "white",
  },
})

const validationSchema = yup.object({
  email: yup.string().required(),
})

export type ConversationAddDialogProps = {
  open: boolean
  onClose: () => void
}

const ConversationAddDialog = ({
  open,
  onClose,
}: ConversationAddDialogProps) => {
  const classes = useStyles()

  return (
    <Dialog open={open} onClose={onClose} classes={{ paper: classes.dialog }}>
      <DialogTitle disableTypography>
        <Typography variant="h5" component="h2">
          Start a conversation
        </Typography>
      </DialogTitle>
      <Formik
        initialValues={{ email: "" }}
        validationSchema={validationSchema}
        onSubmit={async (val, helpers) => {
          helpers.setSubmitting(true)

          let docId: string = ""
          try {
            docId = await addConversation(val.email)
            helpers.resetForm()
          } catch (err) {
            console.error(
              `An error occurred trying to create the conversation`,
              err
            )
          }
          helpers.setSubmitting(false)

          if (docId) {
            onClose()
          }
        }}
      >
        {({ handleBlur, handleChange, isSubmitting, values }) => (
          <Form>
            <DialogContent>
              <Typography variant="body1" component="p">
                Enter the email of a user you would like to chat with:
              </Typography>
              <TextField
                autoFocus
                disabled={isSubmitting}
                inputProps={{
                  "aria-label": "Email",
                }}
                className={classes.input}
                variant="outlined"
                name="email"
                value={values.email}
                onChange={handleChange}
                onBlur={handleBlur}
                inputRef={el => {
                  if (!el) return

                  el.focus()
                }}
              />
            </DialogContent>
            <DialogActions className={classes.inputActions}>
              <Button
                color="default"
                variant="contained"
                disabled={isSubmitting}
                onClick={e => {
                  e.preventDefault()

                  onClose()
                }}
              >
                Cancel
              </Button>
              <Button
                color="primary"
                variant="contained"
                type="submit"
                disabled={isSubmitting || !values.email}
              >
                {!isSubmitting && "Add"}
                {isSubmitting && (
                  <CircularProgress className={classes.spinner} size="1.5rem" />
                )}
              </Button>
            </DialogActions>
          </Form>
        )}
      </Formik>
    </Dialog>
  )
}

export default ConversationAddDialog
