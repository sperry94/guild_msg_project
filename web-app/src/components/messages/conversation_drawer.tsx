import React, { useState } from "react"
import useMediaQuery from "@material-ui/core/useMediaQuery"
import { makeStyles, useTheme } from "@material-ui/core/styles"
import { Button, Drawer, IconButton, Typography } from "@material-ui/core"
import Add from "@material-ui/icons/Add"
import Menu from "@material-ui/icons/Menu"
import ConversationAddDialog from "./conversation_add_dialog"
import ConversationList from "./conversation_list"

const useStyles = makeStyles(theme => ({
  toolbarOffset: {
    ...theme.mixins.toolbar,
  },
  drawer: {
    position: "relative",
    zIndex: 2,
    width: "300px",
    height: "100vh",
  },
  drawerPaper: {
    width: "300px",
    height: "100vh",
  },
  addNewBtn: {
    backgroundColor: theme.palette.grey[900],
    borderRadius: 0,
    borderBottom: "none",
    borderLeft: "none",
    borderRight: "none",
    position: "absolute",
    height: "73px",
    width: "100%",
    bottom: 0,
    left: 0,
    right: 0,
  },
  backBtn: {
    position: "absolute",
    zIndex: 4,
  },
}))

export type ConversationDrawerProps = {
  conversationId?: string
}

const ConversationDrawer = (props: ConversationDrawerProps) => {
  const { conversationId } = props

  const [open, setOpen] = useState(true)

  const classes = useStyles()

  const theme = useTheme()
  const isSmall = useMediaQuery(theme.breakpoints.down("sm"))

  const [addingConversation, setAddingConversation] = useState(false)

  return (
    <>
      {isSmall && (
        <div className={classes.backBtn}>
          <IconButton onClick={() => setOpen(true)}>
            <Menu />
          </IconButton>
        </div>
      )}
      <Drawer
        className={classes.drawer}
        variant={isSmall ? "temporary" : "persistent"}
        anchor="left"
        open={!isSmall || open}
        onClose={() => setOpen(false)}
        classes={{
          paper: classes.drawerPaper,
        }}
      >
        {!isSmall && <div className={classes.toolbarOffset} />}
        <ConversationList
          conversationId={conversationId}
          onItemSelect={() => setOpen(false)}
        />

        <Button
          className={classes.addNewBtn}
          variant="outlined"
          onClick={() => setAddingConversation(true)}
        >
          <Add />
          <Typography>New Conversation</Typography>
        </Button>

        <ConversationAddDialog
          open={addingConversation}
          onClose={() => setAddingConversation(false)}
        />
      </Drawer>
    </>
  )
}

export default ConversationDrawer
