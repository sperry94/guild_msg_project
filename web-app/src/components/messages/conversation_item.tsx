import React from "react"
import { Link } from "gatsby"
import { ListItem, Typography, makeStyles, Avatar } from "@material-ui/core"
import { Conversation } from "../../services/conversations"

const useStyles = makeStyles({
  link: {
    textDecoration: "none",
    width: "100%",
    height: "100%",
  },
  avatar: {
    color: "white",
    marginRight: "8px",
  },
})

export type ConversationItemProps = {
  conversationId?: string
  conversation: Conversation
  onSelect?: () => void
}

const ConversationItem = (props: ConversationItemProps) => {
  const { conversationId, conversation, onSelect = () => undefined } = props

  const classes = useStyles()

  return (
    <li>
      <ListItem
        className={classes.link}
        selected={conversationId === conversation.id}
        button
        divider
        component={Link}
        to={`/messages/${conversation.id}`}
        onClick={onSelect}
      >
        {conversation.users.length > 0 && (
          <Avatar
            className={classes.avatar}
            alt={conversation.users[0].displayName}
            src={conversation.users[0].photoUrl}
          >
            {conversation.users[0].displayName.substr(0, 1)}
          </Avatar>
        )}
        <Typography color="inherit">
          {conversation.users.map(u => u.displayName).join(", ")}
        </Typography>
      </ListItem>
    </li>
  )
}

export default ConversationItem
