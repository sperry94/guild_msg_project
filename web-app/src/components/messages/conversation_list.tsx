import React from "react"
import { makeStyles } from "@material-ui/core/styles"
import { CircularProgress, List, Typography } from "@material-ui/core"
import { useConversationList } from "../../services/conversations"
import ConversationItem from "./conversation_item"

const useStyles = makeStyles({
  nav: {
    height: "calc(100% - 136px)",
    overflowY: "auto",
  },
  noConversationText: {
    textAlign: "center",
    marginTop: "48px",
  },
  spinner: {
    display: "block",
    margin: "48px auto",
    color: "white",
  },
})

export type ConversationListProps = {
  conversationId?: string
}

const ConversationList = (props: ConversationListProps) => {
  const { conversationId } = props

  const classes = useStyles()

  const conversations = useConversationList()

  if (!conversations) {
    return <CircularProgress className={classes.spinner} />
  }

  if (conversations.length === 0) {
    return (
      <Typography
        className={classes.noConversationText}
        variant="h6"
        component="p"
      >
        Add a conversation below
      </Typography>
    )
  }

  return (
    <nav className={classes.nav}>
      <List>
        {conversations.map(conversation => (
          <ConversationItem
            key={conversation.id}
            conversationId={conversationId}
            conversation={conversation}
          />
        ))}
      </List>
    </nav>
  )
}

export default ConversationList
