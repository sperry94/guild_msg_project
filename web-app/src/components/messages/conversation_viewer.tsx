import React from "react"
import cn from "classnames"
import { makeStyles } from "@material-ui/core/styles"
import MessageInput from "./message_input"
import MessageList from "./message_list"

const useStyles = makeStyles({
  content: {
    position: "relative",
  },
})

export type ConversationViewerProps = {
  className?: string
  conversationId?: string
  onBackPressed?: () => undefined
}

const ConversationViewer = (props: ConversationViewerProps) => {
  const { className, conversationId } = props

  const classes = useStyles()

  return (
    <div className={cn(className, classes.content)}>
      <MessageList conversationId={conversationId} />
      <MessageInput
        disabled={!conversationId}
        conversationId={conversationId}
      />
    </div>
  )
}

export default ConversationViewer
