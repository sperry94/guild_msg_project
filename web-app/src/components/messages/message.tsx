import React from "react"
import cn from "classnames"
import { firestore } from "firebase/app"
import { makeStyles } from "@material-ui/core/styles"
import { Paper, Typography } from "@material-ui/core"

const useStyles = makeStyles(theme => ({
  outerContent: {
    display: "flex",
    flexDirection: "column",
    "& + &": {
      marginTop: "48px",
    },
    maxWidth: "calc(100% - 48px)",
  },
  content: {
    padding: "24px",
    flex: 0,
  },
  outerFromCurrentUser: {
    alignItems: "end",
    marginLeft: "auto",
    marginRight: "24px",
    [theme.breakpoints.down("sm")]: {
      marginRight: "0px",
    },
  },
  outerFromOtherUser: {
    alignItems: "start",
    marginRight: "auto",
    marginLeft: "24px",
    [theme.breakpoints.down("sm")]: {
      marginLeft: "0px",
    },
  },
  fromCurrentUser: {
    backgroundColor: theme.palette.primary.dark,
  },
  fromOtherUser: {
    backgroundColor: theme.palette.secondary.dark,
  },
  time: {
    width: "100%",
    color: theme.palette.grey[400],
  },
  timeFromCurrentUser: {
    textAlign: "right",
  },
  timeFromOtherUser: {
    textAlign: "left",
  },
}))

export type MessageProps = {
  fromCurrentUser: boolean
  time: Date
  message: string
}

const Message = (props: MessageProps) => {
  const { fromCurrentUser, message, time } = props

  const classes = useStyles()

  return (
    <div
      className={cn(classes.outerContent, {
        [classes.outerFromCurrentUser]: fromCurrentUser,
        [classes.outerFromOtherUser]: !fromCurrentUser,
      })}
    >
      <Typography
        className={cn(classes.time, {
          [classes.timeFromCurrentUser]: fromCurrentUser,
          [classes.timeFromOtherUser]: !fromCurrentUser,
        })}
        variant="caption"
      >
        {time.toLocaleString()}
      </Typography>
      <Paper
        className={cn(classes.content, {
          [classes.fromCurrentUser]: fromCurrentUser,
          [classes.fromOtherUser]: !fromCurrentUser,
        })}
        elevation={3}
      >
        <Typography>{message}</Typography>
      </Paper>
    </div>
  )
}

export default Message
