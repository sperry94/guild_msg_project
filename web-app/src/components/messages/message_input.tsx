import React, { useEffect, useRef, useState, useLayoutEffect } from "react"
import { Formik, Form } from "formik"
import * as yup from "yup"
import { makeStyles } from "@material-ui/core/styles"
import { Button, CircularProgress, Paper, TextField } from "@material-ui/core"
import { sendMessage } from "../../services/messages"

const useStyles = makeStyles(theme => ({
  content: {
    position: "absolute",
    backgroundColor: theme.palette.grey[900],
    width: "100%",
    minHeight: "72px",
    padding: "8px 24px",
    bottom: 0,
    left: 0,
    right: 0,
  },
  form: {
    display: "flex",
    alignItems: "center",
    minHeight: "52px",
  },
  input: {
    flex: 1,
    marginRight: "24px",
  },
  spinner: {
    color: "white",
  },
}))

const validationSchema = yup.object({
  message: yup.string().required(),
})

export type MessageFormData = {
  message: string
}

export type MessageInputProps = {
  conversationId: string | undefined
  disabled?: boolean
}

const MessageInput = (props: MessageInputProps) => {
  const { disabled = false, conversationId } = props

  const classes = useStyles()

  const resetFormRef = useRef<null | (() => void)>(null)

  useEffect(() => {
    const resetFn = resetFormRef.current || (() => undefined)
    resetFn()
  }, [conversationId])

  return (
    <Paper className={classes.content} elevation={5} square>
      <Formik
        enableReinitialize
        initialValues={{ message: "" }}
        validationSchema={validationSchema}
        onSubmit={async (val, helpers) => {
          if (!conversationId) return

          helpers.resetForm()
          helpers.setSubmitting(true)

          try {
            await sendMessage(conversationId, val.message)
          } catch (err) {
            console.error(`An error occurred trying to send the message`, err)
            helpers.setFieldValue("message", val.message)
          }

          helpers.setSubmitting(false)
        }}
      >
        {({ handleBlur, handleChange, isSubmitting, resetForm, values }) => {
          useEffect(() => {
            resetFormRef.current = resetFormRef.current || resetForm
          }, [])

          return (
            <Form className={classes.form}>
              <TextField
                autoFocus
                disabled={disabled || isSubmitting}
                inputProps={{
                  "aria-label": "Message",
                }}
                className={classes.input}
                variant="outlined"
                name="message"
                value={values.message}
                onChange={handleChange}
                onBlur={handleBlur}
                inputRef={el => {
                  if (!el) return

                  el.focus()
                }}
              />
              <Button
                disabled={disabled || isSubmitting || !values.message}
                type="submit"
                color="primary"
                variant="contained"
              >
                {!isSubmitting && "Send"}
                {isSubmitting && (
                  <CircularProgress className={classes.spinner} size="1.5rem" />
                )}
              </Button>
            </Form>
          )
        }}
      </Formik>
    </Paper>
  )
}

export default MessageInput
