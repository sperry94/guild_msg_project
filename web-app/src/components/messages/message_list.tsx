import React, { useLayoutEffect, useRef } from "react"
import { makeStyles } from "@material-ui/core/styles"
import { CircularProgress, Typography } from "@material-ui/core"
import { useAuthContext } from "../../services/auth"
import { useMessages } from "../../services/messages"
import Message from "./message"

const useStyles = makeStyles({
  content: {
    height: "calc(100% - 72px)",
    overflowY: "auto",
  },
  innerContent: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "end",
    padding: "24px",
    minHeight: "100%",
  },
  noConversationText: {
    textAlign: "center",
    marginTop: "48px",
  },
  spinner: {
    display: "block",
    margin: "48px auto",
    color: "white",
  },
})

export type MessageListProps = {
  className?: string
  conversationId?: string
}

const MessageList = (props: MessageListProps) => {
  const { conversationId } = props

  const listRef = useRef<HTMLDivElement | null>(null)

  const classes = useStyles()

  const auth = useAuthContext()

  const messages = useMessages(conversationId)

  useLayoutEffect(() => {
    if (!messages || !listRef.current) return

    listRef.current.scrollTop = listRef.current.scrollHeight
  }, [messages, listRef.current])

  if (!auth) {
    return null
  }

  if (!conversationId) {
    return (
      <Typography
        className={classes.noConversationText}
        variant="h6"
        component="p"
      >
        Please select a conversation from the list on the left or start a new
        one!
      </Typography>
    )
  }

  if (!messages) {
    return <CircularProgress className={classes.spinner} />
  }

  if (messages.length === 0) {
    return (
      <Typography
        className={classes.noConversationText}
        variant="h6"
        component="p"
      >
        Send a message!
      </Typography>
    )
  }

  return (
    <div className={classes.content} ref={lRef => (listRef.current = lRef)}>
      <div className={classes.innerContent}>
        {messages.map((msg, index) => (
          <Message
            key={index}
            fromCurrentUser={msg.senderId === auth.uid}
            time={msg.time}
            message={msg.message}
          />
        ))}
      </div>
    </div>
  )
}

export default MessageList
