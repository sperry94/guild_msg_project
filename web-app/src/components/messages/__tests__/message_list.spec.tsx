import React from "react"
import { useAuthContext } from "../../../services/auth"
import { useMessages, Message } from "../../../services/messages"
import renderer from "react-test-renderer"
import MessageList from "../message_list"

jest.mock("../../../services/auth")
jest.mock("../../../services/messages")

const mocked_useAuthContext = useAuthContext as jest.Mock<any>
const mocked_useMessages = useMessages as jest.Mock<Message[] | null>

describe("Message List", () => {
  it(`renders null when no auth`, () => {
    mocked_useAuthContext.mockReturnValueOnce(null)
    mocked_useMessages.mockReturnValueOnce([
      {
        senderId: "sender-id",
        time: new Date(2019, 1, 3, 3, 45, 12),
        message: "hello",
      },
    ])
    const tree = renderer.create(<MessageList />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it(`renders message when no conversation id`, () => {
    mocked_useAuthContext.mockReturnValueOnce({})
    mocked_useMessages.mockReturnValueOnce([
      {
        senderId: "sender-id",
        time: new Date(2019, 1, 3, 3, 45, 12),
        message: "hello",
      },
    ])
    const tree = renderer.create(<MessageList />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it(`renders spinner when lookup up messages`, () => {
    mocked_useAuthContext.mockReturnValueOnce({})
    mocked_useMessages.mockReturnValueOnce(null)
    const tree = renderer
      .create(<MessageList conversationId="convo-id" />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it(`renders message when no messages`, () => {
    mocked_useAuthContext.mockReturnValueOnce({})
    mocked_useMessages.mockReturnValueOnce([])
    const tree = renderer
      .create(<MessageList conversationId="convo-id" />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it(`renders messages`, () => {
    mocked_useAuthContext.mockReturnValueOnce({ uid: "me" })
    mocked_useMessages.mockReturnValueOnce([
      {
        senderId: "me",
        time: new Date(2019, 4, 3, 3, 45, 12),
        message: "hello",
      },
      {
        senderId: "my-friend",
        time: new Date(2019, 1, 3, 3, 45, 12),
        message: "hello",
      },
    ])
    const tree = renderer
      .create(<MessageList conversationId="convo-id" />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
