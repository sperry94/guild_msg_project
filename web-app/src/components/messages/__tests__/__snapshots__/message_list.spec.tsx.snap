// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message List renders message when no conversation id 1`] = `
<p
  className="MuiTypography-root makeStyles-noConversationText-3 MuiTypography-h6"
>
  Please select a conversation from the list on the left or start a new one!
</p>
`;

exports[`Message List renders message when no messages 1`] = `
<p
  className="MuiTypography-root makeStyles-noConversationText-3 MuiTypography-h6"
>
  Send a message!
</p>
`;

exports[`Message List renders messages 1`] = `
<div
  className="makeStyles-content-1"
>
  <div
    className="makeStyles-innerContent-2"
  >
    <div
      className="makeStyles-outerContent-47 makeStyles-outerFromCurrentUser-49"
    >
      <span
        className="MuiTypography-root makeStyles-time-53 makeStyles-timeFromCurrentUser-54 MuiTypography-caption"
      >
        5/3/2019, 3:45:12 AM
      </span>
      <div
        className="MuiPaper-root makeStyles-content-48 makeStyles-fromCurrentUser-51 MuiPaper-elevation3 MuiPaper-rounded"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          hello
        </p>
      </div>
    </div>
    <div
      className="makeStyles-outerContent-47 makeStyles-outerFromOtherUser-50"
    >
      <span
        className="MuiTypography-root makeStyles-time-53 makeStyles-timeFromOtherUser-55 MuiTypography-caption"
      >
        2/3/2019, 3:45:12 AM
      </span>
      <div
        className="MuiPaper-root makeStyles-content-48 makeStyles-fromOtherUser-52 MuiPaper-elevation3 MuiPaper-rounded"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          hello
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Message List renders null when no auth 1`] = `null`;

exports[`Message List renders spinner when lookup up messages 1`] = `
<div
  className="MuiCircularProgress-root makeStyles-spinner-4 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
  role="progressbar"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
>
  <svg
    className="MuiCircularProgress-svg"
    viewBox="22 22 44 44"
  >
    <circle
      className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
      cx={44}
      cy={44}
      fill="none"
      r={20.2}
      strokeWidth={3.6}
      style={Object {}}
    />
  </svg>
</div>
`;
